bpf/lib/nodeport.h

```diff
@@ -192,7 +192,8 @@ static __always_inline int nodeport_nat_ipv6_fwd(struct __ctx_buff *ctx,
 		ret = CTX_ACT_OK;
 
 	/* See the equivalent v4 path for comment */
-	bpf_mark_snat_done(ctx);
+	if (!IS_ERR(ret)) 
```

Comment:
Just realized with the v1.13 backport that this should be before the NAT_PUNT_TO_STACK check like in the ipv4 case.

Reply:
Fixed.

Reply:
Here's what I cooked up:

```
diff --git a/bpf/lib/nodeport.h b/bpf/lib/nodeport.h
index fddd37425a..58d597d96c 100644
--- a/bpf/lib/nodeport.h
+++ b/bpf/lib/nodeport.h
@@ -182,18 +182,21 @@ static __always_inline int nodeport_nat_ipv6_fwd(struct __ctx_buff *ctx,
                .min_port = NODEPORT_PORT_MIN_NAT,
                .max_port = NODEPORT_PORT_MAX_NAT,
        };
-       int ret;
+       int ret = CTX_ACT_OK;
 
        ipv6_addr_copy(&target.addr, addr);
 
-       ret = snat_v6_needed(ctx, addr) ?
-             snat_v6_process(ctx, NAT_DIR_EGRESS, &target) : CTX_ACT_OK;
+       if (snat_v6_needed(ctx, addr)) {
+               ret = snat_v6_process(ctx, NAT_DIR_EGRESS, &target);
+
+               /* See the equivalent v4 path for comment */
+               if (!IS_ERR(ret))
+                       bpf_mark_snat_done(ctx);
+       }
+
        if (ret == NAT_PUNT_TO_STACK)
                ret = CTX_ACT_OK;
 
-       /* See the equivalent v4 path for comment */
-       bpf_mark_snat_done(ctx);
-
        return ret;
 }
 
@@ -1112,17 +1115,20 @@ static __always_inline int nodeport_nat_ipv4_fwd(struct __ctx_buff *ctx)
        bool snat_needed;
 
        snat_needed = snat_v4_prepare_state(ctx, &target);
-       if (snat_needed)
+       if (snat_needed) {
                ret = snat_v4_process(ctx, NAT_DIR_EGRESS, &target);
+
+               /* If multiple netdevs process an outgoing packet, then this packets will
+                * be handled multiple times by the "to-netdev" section. This can lead
+                * to multiple SNATs. To prevent from that, set the SNAT done flag.
+                */
+               if (!IS_ERR(ret))
+                       bpf_mark_snat_done(ctx);
+       }
+
        if (ret == NAT_PUNT_TO_STACK)
                ret = CTX_ACT_OK;
 
-       /* If multiple netdevs process an outgoing packet, then this packets will
-        * be handled multiple times by the "to-netdev" section. This can lead
-        * to multiple SNATs. To prevent from that, set the SNAT done flag.
-        */
-       bpf_mark_snat_done(ctx);
-
        return ret;
 }
```

Reply:
Applied.